### PR TITLE
Adding more pages

### PR DESCRIPTION
**Pull request by robotenique**
Hi, I added the pages for GANs, RNNs, NTMs and LSTMs.
I also fixed a typo in the README.md.

### CODE REVIEW

File: jekyll/content/terms/neural-turing-machine-ntm.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ - neural-network
+ - long-short-term-memory
+references:
+ - "[Neural Turing Machine - Wikipedia](hhttps://en.wikipedia.org/wiki/Neural_Turing_machine)"
+ - "[Neural Turing Machines](https://arxiv.org/pdf/1410.5401.pdf)"
+---
+
+Neural Turing Machines (NTM) are a differentiable computer, an consists of a RNN (commonly with LSTM), and a memory bank, where the neural network can make write and read operations. By making each operation of the NTM differentiable, it can be trained efficiently trained with gradient descent. 
+
+The main idea of the NTM is to use the memory bank - a large, adressable memory - to give a memory to the RNN so that it can read and write to, yielding a pratical mechanism to learn programs. The NTM has been shown to be able to infer simple algorithms, such as copying, sorting and associative recall from input and output examples.
```
Comment (jamesmishra) [inline jekyll/content/terms/neural-turing-machine-ntm.md]: s/adressable/addressable/g

File: jekyll/content/terms/neural-turing-machine-ntm.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+related_terms:
+ - recurrent-neural-network
+ - neural-network
+ - long-short-term-memory
+references:
+ - "[Neural Turing Machine - Wikipedia](hhttps://en.wikipedia.org/wiki/Neural_Turing_machine)"
+ - "[Neural Turing Machines](https://arxiv.org/pdf/1410.5401.pdf)"
+---
+
+Neural Turing Machines (NTM) are a differentiable computer, an consists of a RNN (commonly with LSTM), and a memory bank, where the neural network can make write and read operations. By making each operation of the NTM differentiable, it can be trained efficiently trained with gradient descent.
```
Comment (jamesmishra) [inline jekyll/content/terms/neural-turing-machine-ntm.md]: I would say that Neural Turing Machines are a type of recurrent neural network, instead of describing them as a computer.

The [Wikipedia page for Neural Turing Machines](https://en.wikipedia.org/wiki/Neural_Turing_machine) describes them in a similar way.

File: jekyll/content/terms/neural-turing-machine-ntm.md (hunk 1 above)
Comment (jamesmishra) [inline jekyll/content/terms/neural-turing-machine-ntm.md]: Instead of using a single dash  `-`, we should use two dashes `--` because Pandoc will convert into a longer "em" dash `—`.

File: jekyll/content/terms/recurrent-neural-network.md [hunk 3]
```diff
@@ -3,4 +3,12 @@ title: Recurrent Neural Network
 related_terms:
  - recursive-neural-network
  - neural-network
+ - long-short-term-memory
+references:
+ - "[Recurrent neural networks - Wikipedia](https://en.wikipedia.org/wiki/Recurrent_neural_network)"
+ - "[Recurrent neural networks](https://crl.ucsd.edu/~elman/Papers/fsit.pdf)"
 ---
+
+Recurrent neural networks (RNN) are feed-forward neural networks, but aren't stateless. An RNN have an internal loop that allows information to persist in the network. Neurons receive information not just from the previous layer, but also from themselves from the previous pass. This means that the order of inputs to the RNN matter, and may give different results with different order, as the RNNs have state.
```
Comment (jamesmishra) [inline jekyll/content/terms/recurrent-neural-network.md]: Instead of saying that RNNs aren't stateless, we should say they contain an internal memory, and this makes them different from traditional feed-forward neural networks.

File: jekyll/content/terms/recurrent-neural-network.md [hunk 4]
```diff
@@ -3,4 +3,12 @@ title: Recurrent Neural Network
...
+
+RNNs are known for the vanishing or exploding gradient, where depending on the activation functions used, the information can get lost over time. Long short term memory (LSTMs) adresses this problem. RNNs are commonly used with data that can be represented as a sequence, due to the fact that an RNN is sensitive to the order of the inputs, like in the natural language processing area.
```
Comment (jamesmishra) [inline jekyll/content/terms/recurrent-neural-network.md]: Let's hyperlink to the LSTM article.

e.g.

```
[Long short-term memory networks][1] address...
```

and then at the bottom

```
[1]: /terms/terms/long-short-term-memory-lstm/
```

Comment (jamesmishra) [inline jekyll/content/terms/recurrent-neural-network.md]: s/adress/addresss/g

Comment (jamesmishra) [inline jekyll/content/terms/recurrent-neural-network.md]: It would be good to mention examples of sequential data... e.g./ text/language, audio video, sensor data collected over time, etc.

File: jekyll/content/terms/generative-adversarial-network-gan.md [hunk 5]
```diff
@@ -5,3 +5,9 @@ references:
  - "[Generative adversarial networks](https://arxiv.org/abs/1406.2661)"
  - "[NIPS 2016 tutorial: Generative Adversarial Networks](https://arxiv.org/abs/1701.00160)"
 ---
+Generative adversarial networks were proposed by Ian Goodfellow in 2014. A GAN consists of a model with two neural networks, G (generative model) and D (discriminative model), which are competing in a minimax two-player game.
```
Comment (jamesmishra) [inline jekyll/content/terms/generative-adversarial-network-gan.md]: GANs are a tricky topic. I avoided filling in this definition because it is such a difficult topic. :)

But I think a really good definition for GANs will highlight on motivations:

 - Why do GANs matter? We use them for certain generative tasks like image completion and superscalar resolution. [More details here](https://www.cs.toronto.edu/~duvenaud/courses/csc2541/slides/gan-applications.pdf).

The core of GANs is that they are two neural networks... jointly trained... with a loss function that represents a minimax game between the two networks.

I think it's a difficult definition to write, and it might make sense to make the GAN definition into a separate pull request if you're interested. I'm happy to help you write it.

Comment (jamesmishra) [reply jekyll/content/terms/generative-adversarial-network-gan.md]: Some improvements and suggestions I have for this article is to:

 - Include and explain the loss function from [the original GAN paper](https://arxiv.org/abs/1406.2661). This might be a little bit much... since we don't do that for other definitions, but when I first started studying GANs, I didn't understand them until I broke down the loss function.

 - Figure out how to describe more clearly that the generator network generates fakes and samples from the training distribution, and the discriminator's job is to tell them apart. You kinda do this... but I'm struggling with a way to rephrase it right now. I know, that's a lame criticism. :)

File: jekyll/content/terms/generative-adversarial-network-gan.md [hunk 6]
```diff
@@ -5,3 +5,9 @@ references:
...
+
+The neural network G is modeled to capture the data distribution, and it's purpose is to generate new content. The network D is the discriminator, which purpose is to estimate the probability that a sample came from the training data rather then generated by G.
+The training of the G network ultimately is to maximize the probability of D making a mistake (minimax strategy). The mutual training of the GAN creates a form of competition where D is getting better at the prediction (distinguishing between real and generated data), and G is learning to become less predictable.
+
+GANs are difficult to train due to the fact that the training of G and D must be well synchronized. If D or G becomes too good compared to the other, a GAN won't converge (Helvetica scenario). GANs don't need to use Markov chains and no inference on the data is needed in the learning sessions.
```
Comment (jamesmishra) [inline jekyll/content/terms/generative-adversarial-network-gan.md]: I think it is worth mentioning mode collapse / the Helvetica scenario, but I wouldn't say that "GANs are difficult to train" because that statement will become out-of-date as soon as someone publishes a paper that makes GANs easy to train.

And then we would be wrong! :)

File: jekyll/content/terms/generative-adversarial-network-gan.md (hunk 5 above)
Comment (jamesmishra) [reply jekyll/content/terms/generative-adversarial-network-gan.md]: I know that's a lot of feedback, but this is also why I haven't bothered filling out this definition either...

File: jekyll/content/terms/recurrent-neural-network.md (hunk 4 above)
Comment (jamesmishra) [inline jekyll/content/terms/recurrent-neural-network.md]: I would say something like, "RNNs are particularly sensitive to the vanishing gradient and exploding gradient problems."

**Review by jamesmishra — CHANGES_REQUESTED** [review]
Hey, thanks for being the first person to submit a pull request for Machine Learning Glossary.

You'll go down in history as one of the core authors of this website. :)

I think you've done a great job filling out some important articles, but I do have some feedback before we publish all this on the main website.

If you want to split the GAN stuff into a separate pull request, or just don't want to worry about it, feel free to do so.

**Comment by robotenique** [comment]
Thank you for your time providing feedback!
I reviewed and added the changes you suggested. For the GAN entry I think it's better to make another pull request.
How should I proceed with the GAN file? Reset it to the master branch and in this pull request, then continue to work on it for the other PR?

**Comment by jamesmishra** [comment]
Yeah, just revert the GAN file to master in this PR... but save it so you can put it into a different branch.

After that, I'll merge the request.

**Comment by robotenique** [comment]
Done!

**Comment by jamesmishra** [comment]
Perfect! The Continuous Integration system will upload your changes now.

It might take a little while to see it on https://machinelearning.wtf .. I haven't configured out deployments to automatically invalidate the Amazon Cloudfront cache yet.
